Source/Diagnostics/Diagnostics.cpp

```diff
@@ -105,60 +110,36 @@ Diagnostics::BaseReadParameters ()
-    // - allocate and fill the array m_rho_per_species_index, which contains the
-    //   indices of the species that request to dump rho per species; the indices
-    //   will be then passed to the constructor of RhoFunctor and used in
-    //   RhoFunctor::operator() to get the correct particle container for the given species
-    amrex::Vector<std::string> species_variables;
-    const MultiParticleContainer& mpc = warpx.GetPartContainer();
-    // If <diag_name>.species is not provided, loop over all species to check whether
-    // rho per species is requested
-    if (m_species_names.size() == 0u) m_species_names = mpc.GetSpeciesNames();
-    // ns_dump_rho: number of species that dump rho per species
-    int ns_dump_rho = 0;
-    // Loop over all species
-    for (const auto& species_name : m_species_names) {
-        // Parse input argument <diag_name>.<species_name>.variables
-        amrex::ParmParse pp_sp(m_diag_name + "." + species_name);
-        if (pp_sp.queryarr("variables", species_variables)) {
-            for (const auto& var : species_variables) {
-                if (var == "rho") {
-                    // Add string "rho_<species_name>" to m_varnames
-                    m_varnames.push_back("rho_" + species_name);
-                    // Count number of species that dump rho per species
-                    ns_dump_rho++;
-                }
-            }
-        }
-        // Clear content of species_variables before moving to the next species
-        species_variables.clear();
-    }
-    // Allocate array of species indices that dump rho per species
-    m_rho_per_species_index.resize(ns_dump_rho);
-    // ns: total number of species
-    const int ns = int(m_species_names.size());
-    // is_dump_rho: species index to loop over species that dump rho per species
-    int is_dump_rho = 0;
-    // Loop over all species
-    for (int is = 0; is < ns; is++) {
-        // Parse input argument <diag_name>.<species_name>.variables
-        amrex::ParmParse pp_sp(m_diag_name + "." + m_species_names[is]);
-        if (pp_sp.queryarr("variables", species_variables)) {
-            for (const auto& var : species_variables) {
-                if (var == "rho") {
-                    // Fill array of species indices that dump rho per species
-                    m_rho_per_species_index.at(is_dump_rho) = is;
-                    is_dump_rho++;
+    // Get names of all species in the simulation, independently of the subset
+    // of species requested in <diag_name>.species and stored in m_species_names
+    const std::vector<std::string> species_names = warpx.GetPartContainer().GetSpeciesNames();
+    // Auxiliary variables
+    std::string species;
+    bool species_name_is_wrong;
+    // Loop over all fields stored in m_varnames
+    for (const auto& var : m_varnames) {
+        // Check if m_varnames contains a string of the form rho_<species_name>
+        if (var.find("rho_") != std::string::npos) {
```

Comment:
It looks like this will return true if the string contains `rho_` even if it does not start by `rho_`, am I correct?

Do we actually want to do that? For instance if the user makes a typo and forgets a space you could have `diag.fields_to_plot = Ex Ez Byrho_electron`. This would not crash, output `Ex`, `Ez` and `rho_electron`, and ignore `By`, which would not be ideal right?
(Note that if you change this you will also need to change it in `FullDiagnostics.cpp`)

On a similar note (but not really about this PR), I've just tested and if the user makes a typo in `field_to_plot` we currently get a Segfault instead of a clear error message. :grimacing: I'll write a mini-PR to fix that.

Reply:
You could try something like `var.rfind("rho_", 0) == 0` ...